auth/src/main/kotlin/com/tidal/sdk/auth/login/LoginRepository.kt

```diff
@@ -78,6 +80,7 @@ internal class LoginRepository constructor(
     }
 
     suspend fun setCredentials(credentials: Credentials, refreshToken: String? = null) {
+        networkingJobHandler.cancelAllJobs()
```

Comment:
I think there's a problem with this. You're going to cancel getCredentials jobs that have been added to the collection before getting here, yes. But that doesn't cover jobs that are added after this line is executed. Those will be capable of overwriting the credentials set by this method, even if the corresponding call to getCredentials happened before. You would need to run job cancellation within the mutex too I think, so that you achieve what my understanding of what you want is - have token retrieval and write run exclusively with each other. But that will give you a new problem - if there's a write request while ownership of the mutex is held by a retrieval request, it needs to bypass the wait by cancelling the existing ownership of the mutex. That may be tricky to achieve with this approach I think.

I think the easiest approach to do what you need is to rework concurrency management in the sdk to use [Handler](https://developer.android.com/reference/android/os/Handler). It's a queue where you can send runnables. It's backed by a single thread, so it's thread-safe by default. You'll have two types of runnables - one for getCredentials and one for setCredentials. If you receive a request for setCredentials, you add the runnable to the end of the queue - if it's for setCredentials, you do removeCallbacks for your currently queued setCredentials Runnable if any, save the new Runnable as your current one so you can remove it later, and post it at the front of the queue so it runs before any other getCredentials requests.

Thoughts?

Reply:
Well, mainly I am trying to pragmatically present a solution for a specific issue that occurs only when using our "backdoor" `setCredentials` (and only when using during certain UI tests). Both my attempts solve the problem (I verified) without causing new issues elsewhere.

The other approach avoided the problems you suspect this one here has, by choosing a less "correct" approach than this here. For this edge case that is perfectly fine imho. 

It's fine to disagree with that though, and since you probably have had more exposure to coroutine concurrency management in the player, and appear you know how to solve this issue right, would you mind creating a proposal?

Reply:
There also is the option to pick one solution, and creating a task to improve the situation in a more general approach, which might be the most pragmatic.